File: Sources/KakaoSapiens/Services/SyncPairingClient.swift
import Foundation

/// The four pairing calls, and nothing else.
///
/// Two of them are authenticated as the host device and two are not: a joiner
/// has no token yet, which is the whole reason pairing exists. The token is
/// read per request rather than captured, for the same reason the pull client
/// reads it per request — a client built before enrollment would otherwise
/// stay unauthenticated for the rest of the session.
public struct SyncPairingClient {
    public struct Claim: Equatable {
        public let claimID: String
        public let claimLookup: String
        public let claimEnvelope: String
        public let state: String
    }

    public struct Session: Equatable {
        public let sessionID: String
        public let expiresAt: String
        public let replayed: Bool
    }

    public enum PairingClientError: Error, Equatable {
        /// The status alone. Server text could carry a lookup or an envelope.
        case httpStatus(Int)
        case malformedResponse
        case notAuthenticated
    }

    private let baseURL: URL
    private let token: () -> Data?
    private let transport: SyncHTTPTransport

    public init(baseURL: URL, token: @escaping () -> Data?, transport: SyncHTTPTransport) {
        self.baseURL = baseURL
        self.token = token
        self.transport = transport
    }

    // MARK: host

    public func createSession(sessionID: String, sessionLookup: Data) async throws -> Session {
        let result = try await send(
            path: "/v1/pairing/sessions",
            body: [
                "protocol_version": 1,
                "session_id": sessionID,
                "pairing_session_lookup": Self.wire(sessionLookup),
            ],
            authenticated: true
        )
        guard let id = result["session_id"] as? String,
              let expires = result["expires_at"] as? String,
              let status = result["status"] as? String else {
            throw PairingClientError.malformedResponse
        }
        return Session(sessionID: id, expiresAt: expires, replayed: status == "replayed")
    }

    public func listClaims(sessionID: String) async throws -> [Claim] {
        let result = try await send(
            path: "/v1/pairing/sessions/\(sessionID)/claims",
            body: nil,
            authenticated: true
        )
        guard let raw = result["claims"] as? [[String: Any]] else {
            throw PairingClientError.malformedResponse
        }
        return try raw.map { entry in
            guard let id = entry["claim_id"] as? String,
                  let lookup = entry["claim_lookup"] as? String,
                  let envelope = entry["claim_envelope"] as? String,
                  let state = entry["state"] as? String else {
                throw PairingClientError.malformedResponse
            }
            return Claim(claimID: id, claimLookup: lookup, claimEnvelope: envelope, state: state)
        }
    }

    public func approve(
        sessionID: String,
        claimID: String,
        claimLookup: Data,
        deliveryEnvelope: Data,
        device: (id: String, spaceID: String, platform: String, tokenHash: String)
    ) async throws {
        _ = try await send(
            path: "/v1/pairing/sessions/\(sessionID)/claims/\(claimID)/approve",
            body: [
                "protocol_version": 1,
                "claim_lookup": Self.wire(claimLookup),
                "delivery_envelope": Self.wire(deliveryEnvelope),
                "device": [
                    "device_id": device.id,
                    "space_id": device.spaceID,
                    "platform": device.platform,
                    "display_name": NSNull(),
                    "device_token_hash": device.tokenHash,
                ],
            ],
            authenticated: true
        )
    }

    // MARK: joiner

    public func submitClaim(
        sessionID: String,
        sessionLookup: Data,
        claimID: String,
        claimLookup: Data,
        claimEnvelope: Data,
        redeemVerifier: String
    ) async throws {
        let result = try await send(
            path: "/v1/pairing/sessions/\(sessionID)/claims",
            body: [
                "protocol_version": 1,
                "pairing_session_lookup": Self.wire(sessionLookup),
                "claim_id": claimID,
                "claim_lookup": Self.wire(claimLookup),
                "claim_envelope": Self.wire(claimEnvelope),
                "claim_redeem_verifier": redeemVerifier,
            ],
            authenticated: false
        )
        // Deliberately no ciphertext is expected here. At submit time the host
        // has not approved anything, so a response carrying a delivery package
        // would mean the server handed out a package nobody authorised.
        guard result["status"] as? String == "submitted" else {
            throw PairingClientError.malformedResponse
        }
    }

    /// Returns the delivery envelope. Only ever succeeds once per claim.
    public func redeem(
        sessionID: String,
        claimID: String,
        claimLookup: Data,
        redeemAuth: Data
    ) async throws -> Data {
        let result = try await send(
            path: "/v1/pairing/sessions/\(sessionID)/claims/\(claimID)/redeem",
            body: [
                "protocol_version": 1,
                "claim_lookup": Self.wire(claimLookup),
                "claim_redeem_auth": Self.wire(redeemAuth),
            ],
            authenticated: false
        )
        guard let encoded = result["delivery_envelope"] as? String,
              let envelope = Data(base64Encoded: encoded) else {
            throw PairingClientError.malformedResponse
        }
        return envelope
    }

    // MARK: transport

    private func send(
        path: String,
        body: [String: Any]?,
        authenticated: Bool
    ) async throws -> [String: Any] {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = body == nil ? "GET" : "POST"
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
        if authenticated {
            // Refused here rather than sent unauthenticated.
            guard let deviceToken = token(), deviceToken.count == 32 else {
                throw PairingClientError.notAuthenticated
            }
            request.setValue("Device gdt1_\(Self.tokenText(deviceToken))", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [.sortedKeys])
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let response = try await transport.send(request)
        let data = response.body
        guard (200..<300).contains(response.statusCode) else {
            throw PairingClientError.httpStatus(response.statusCode)
        }
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              root["protocol_version"] as? Int == 1,
              let result = root["result"] as? [String: Any] else {
            throw PairingClientError.malformedResponse
        }
        return result
    }

    /// Request bodies use padded standard Base64, which is what the Worker's
    /// canonical-Base64 check accepts.
    private static func wire(_ data: Data) -> String { data.base64EncodedString() }

    /// The Authorization header is the one place using unpadded Base64URL.
    private static func tokenText(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncPairingClient.kt
package com.sapiens.gagaodok.sync

import java.util.Base64
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

/**
 * The four pairing calls, and nothing else.
 *
 * Two of them are authenticated as the host device and two are not: a joiner
 * has no token yet, which is the whole reason pairing exists. The token is read
 * per request rather than captured, for the same reason the pull client reads
 * it per request.
 */
class SyncPairingClient(
    baseUrl: String,
    private val token: () -> ByteArray?,
    private val transport: SyncHttpTransport = OkHttpSyncTransport(),
) {
    data class Claim(
        val claimId: String,
        val claimLookup: String,
        val claimEnvelope: String,
        val state: String,
    )

    data class Session(val sessionId: String, val expiresAt: String, val replayed: Boolean)

    /** The status alone. Server text could carry a lookup or an envelope. */
    class PairingClientException(val statusCode: Int? = null, val notAuthenticated: Boolean = false) : Exception()

    private val baseUrl = baseUrl.removeSuffix("/")

    init {
        require(this.baseUrl.startsWith("https://") && !this.baseUrl.contains('?') && !this.baseUrl.contains('#'))
    }

    // host

    fun createSession(sessionId: String, sessionLookup: ByteArray): Session {
        val result = send(
            "/v1/pairing/sessions",
            JSONObject()
                .put("protocol_version", 1)
                .put("session_id", sessionId)
                .put("pairing_session_lookup", wire(sessionLookup)),
            authenticated = true,
        )
        return Session(
            sessionId = result.getString("session_id"),
            expiresAt = result.getString("expires_at"),
            replayed = result.optString("status") == "replayed",
        )
    }

    fun listClaims(sessionId: String): List<Claim> {
        val result = send("/v1/pairing/sessions/$sessionId/claims", null, authenticated = true)
        val array: JSONArray = result.getJSONArray("claims")
        return (0 until array.length()).map { index ->
            val entry = array.getJSONObject(index)
            Claim(
                claimId = entry.getString("claim_id"),
                claimLookup = entry.getString("claim_lookup"),
                claimEnvelope = entry.getString("claim_envelope"),
                state = entry.getString("state"),
            )
        }
    }

    fun approve(
        sessionId: String,
        claimId: String,
        claimLookup: ByteArray,
        deliveryEnvelope: ByteArray,
        deviceId: String,
        spaceId: String,
        platform: String,
        deviceTokenHash: String,
    ) {
        send(
            "/v1/pairing/sessions/$sessionId/claims/$claimId/approve",
            JSONObject()
                .put("protocol_version", 1)
                .put("claim_lookup", wire(claimLookup))
                .put("delivery_envelope", wire(deliveryEnvelope))
                .put(
                    "device",
                    JSONObject()
                        .put("device_id", deviceId)
                        .put("space_id", spaceId)
                        .put("platform", platform)
                        .put("display_name", JSONObject.NULL)
                        .put("device_token_hash", deviceTokenHash),
                ),
            authenticated = true,
        )
    }

    // joiner

    fun submitClaim(
        sessionId: String,
        sessionLookup: ByteArray,
        claimId: String,
        claimLookup: ByteArray,
        claimEnvelope: ByteArray,
        redeemVerifier: String,
    ) {
        val result = send(
            "/v1/pairing/sessions/$sessionId/claims",
            JSONObject()
                .put("protocol_version", 1)
                .put("pairing_session_lookup", wire(sessionLookup))
                .put("claim_id", claimId)
                .put("claim_lookup", wire(claimLookup))
                .put("claim_envelope", wire(claimEnvelope))
                .put("claim_redeem_verifier", redeemVerifier),
            authenticated = false,
        )
        // Deliberately no ciphertext is expected here. At submit time the host
        // has not approved anything, so a response carrying a delivery package
        // would mean the server handed out a package nobody authorised.
        if (result.optString("status") != "submitted") throw PairingClientException()
    }

    /** Returns the delivery envelope. Only ever succeeds once per claim. */
    fun redeem(
        sessionId: String,
        claimId: String,
        claimLookup: ByteArray,
        redeemAuth: ByteArray,
    ): ByteArray {
        val result = send(
            "/v1/pairing/sessions/$sessionId/claims/$claimId/redeem",
            JSONObject()
                .put("protocol_version", 1)
                .put("claim_lookup", wire(claimLookup))
                .put("claim_redeem_auth", wire(redeemAuth)),
            authenticated = false,
        )
        return runCatching { Base64.getDecoder().decode(result.getString("delivery_envelope")) }
            .getOrElse { throw PairingClientException() }
    }

    private fun send(path: String, body: JSONObject?, authenticated: Boolean): JSONObject {
        val builder = Request.Builder().url(baseUrl + path).header("Cache-Control", "no-cache")
        if (authenticated) {
            // Refused here rather than sent unauthenticated.
            val deviceToken = token()
            if (deviceToken == null || deviceToken.size != 32) {
                throw PairingClientException(notAuthenticated = true)
            }
            builder.header(
                "Authorization",
                "Device gdt1_" + Base64.getUrlEncoder().withoutPadding().encodeToString(deviceToken),
            )
        }
        val request = if (body == null) {
            builder.get().build()
        } else {
            builder.post(body.toString().toByteArray().toRequestBody(JSON)).build()
        }

        val response = transport.send(request)
        if (response.statusCode !in 200..299) throw PairingClientException(response.statusCode)
        val root = runCatching { JSONObject(String(response.body, Charsets.UTF_8)) }
            .getOrElse { throw PairingClientException() }
        if (root.optInt("protocol_version", -1) != 1) throw PairingClientException()
        return root.optJSONObject("result") ?: throw PairingClientException()
    }

    private companion object {
        val JSON = "application/json".toMediaType()

        /**
         * Request bodies use padded standard Base64, which is what the Worker's
         * canonical-Base64 check accepts. The Authorization header is the one
         * place that uses unpadded Base64URL.
         */
        fun wire(value: ByteArray): String = Base64.getEncoder().encodeToString(value)
    }
}
